File: solutions/go/protein-translation/1/protein_translation.go
package protein

import "errors"

var ErrStop = errors.New("err Stop")
var ErrInvalidBase = errors.New("err base")

func FromRNA(rna string) ([]string, error) {
	result := []string{}
	ii := 3
	for i := 0; i < len(rna); i += 3 {
		codon := ""
		for iii := i; iii < ii; iii++ {
			codon += string(rna[iii])
		}
		protein, e := FromCodon(codon)
		if e == ErrStop {
			break
		} else if e == ErrInvalidBase {
			return result, e
		}
		result = append(result, protein)

		ii += 3
	}
	return result, nil
}

func FromCodon(codon string) (string, error) {
	result := ""
	switch codon {
	/*		Codon                 | Protein
			:---                  | :---
			AUG                   | Methionine
			UUU, UUC              | Phenylalanine
			UUA, UUG              | Leucine
			UCU, UCC, UCA, UCG    | Serine
			UAU, UAC              | Tyrosine
			UGU, UGC              | Cysteine
			UGG                   | Tryptophan
			UAA, UAG, UGA         | STOP
	*/
	case "AUG":
		result = "Methionine"
	case "UUU", "UUC":
		result = "Phenylalanine"
	case "UCU", "UCC", "UCA", "UCG":
		result = "Serine"
	case "UAU", "UAC":
		result = "Tyrosine"
	case "UGU", "UGC":
		result = "Cysteine"
	case "UGG":
		result = "Tryptophan"
	case "UAA", "UAG", "UGA":
		return "", ErrStop
	case "UUA", "UUG":
		result = "Leucine"
	default:
		return "", ErrInvalidBase
	}

	return result, nil
}

File: solutions/lua/protein-translation/1/protein-translation.lua
local function translate_codon(codon)
  local mapping <const> = {
    AUG = "Methionine",
    UUU = "Phenylalanine",
    UUC = "Phenylalanine",
    UUA = "Leucine",
    UUG = "Leucine",
    UCC = "Serine",
    UCA = "Serine",
    UCG = "Serine",
    UCU = "Serine",
    UAU = "Tyrosine",
    UAC = "Tyrosine",
    UGU = "Cysteine",
    UGC = "Cysteine",
    UGG = "Tryptophan",
    UAA = "STOP",
    UAG = "STOP",
    UGA = "STOP"
  }

  if (mapping[codon] == nil) then
    error()
  end
  return mapping[codon]
end

local function translate_rna_strand(rna_strand)
  local result = {}
  for i = 1, #rna_strand, 3 do
    local tmp = string.sub(rna_strand, i, i + 2)
    tmp = translate_codon(tmp)
    if tmp == "STOP" then
      return result
    end
    table.insert(result, tmp)
  end
  return result
end

return {
  codon = translate_codon,
  rna_strand = translate_rna_strand
}
